### ISSUE TITLE
How to get novel variants by vg call ? Or what's the meaning of vg augment?

### ISSUE THREAD

**Issue by JD12138** [vgteam/vg#3567]
Hi,
I have vcf files and fastq files for 80 samples. I firstly construct a vg graph with the 80 vcf files. Secondly I map the fastq to the vg graph. Thirdly, I call variants with the 80 samples. But I didn't get novel variants. The variants are all included in the vcf files that I used to construct the vg graph. I don't know how can I get novel variants? It seems that I just genotype the variants again.（I use the same set of samples to construct graph and call variants by vg) Thank you!
Followings are my code:

**vg giraffe -t 20  \
 -Z ${index}pan_index.giraffe.gbz \
 -m ${index}pan_index.min \
 -d ${index}pan_index.dist \
 -f ${sample}_combined_R1.fastq -f ${sample}_combined_R2.fastq \
 -N ${sample} \
 >${out}${sample}.gam

vg pack -x $xg  -g ${path}${sample}.gam -Q 5 -s 5 -o ${out}${sample}.pack
vg call  $xg -k ${out}${sample}.pack -a  >${out}${sample}.vcf**

**Comment by adamnovak** [comment]
Most variant callers take the linear reference and the reads, and then internally generate sets of candidate variants, evaluate support for them, and make calls.

With vg, we have split this into two tools, since the graph already stores a set of candidate variants. The [wiki guide here](https://github.com/vgteam/vg/wiki/Basic-Operations#calling-variants-with-vg-call) shows how you can use `vg augment` to add candidate variants from the reads into the graph (although that pipeline doesn't feature `vg pack`). The [README also shows how to use `vg augment`](https://github.com/vgteam/vg/blob/01323ce3ebd59a3aceec628306774b8f1b9015b0/README.md#augmentation); it comes before `vg pack` and produces a graph that includes novel as well as known variants.

`vg call` itself doesn't find new candidate variants not in the graph; it just produces calls about the variants that are there already.

**Comment by zhanghaoipp** [comment]
> Most variant callers take the linear reference and the reads, and then internally generate sets of candidate variants, evaluate support for them, and make calls.
> 
> With vg, we have split this into two tools, since the graph already stores a set of candidate variants. The [wiki guide here](https://github.com/vgteam/vg/wiki/Basic-Operations#calling-variants-with-vg-call) shows how you can use `vg augment` to add candidate variants from the reads into the graph (although that pipeline doesn't feature `vg pack`). The [README also shows how to use `vg augment`](https://github.com/vgteam/vg/blob/01323ce3ebd59a3aceec628306774b8f1b9015b0/README.md#augmentation); it comes before `vg pack` and produces a graph that includes novel as well as known variants.
> 
> `vg call` itself doesn't find new candidate variants not in the graph; it just produces calls about the variants that are there already.

Hi adamnovak, 
For this question, JD12138 use 80 samples construct the vg graph, so can he get novel SVs with the reads of these 80 samples? I think if he use reads of some new samples, maybe he can get noval SVs, but not with the same sample set. Am I right?

**Comment by adamnovak** [comment]
I don't quite understand the question @zhanghaoipp.

SVs that are in the input VCFs will be in the graph, and will then be callable with vg, even if they are "novel" in the sense that nobody else knew about them before they went into these VCFs. SVs that are only in the reads that are being mapped to the graph, and that aren't in the graph to begin with, won't be called. You *could* call an SV from the reads if the reads were long enough to just map over a 75 bp deletion and pay for its score cost, but they probably won't be. We don't do any local assembly or anything to try and recover novel SVs.

Does that help?

**Comment by zhanghaoipp** [comment]
> I don't quite understand the question @zhanghaoipp.
> 
> SVs that are in the input VCFs will be in the graph, and will then be callable with vg, even if they are "novel" in the sense that nobody else knew about them before they went into these VCFs. SVs that are only in the reads that are being mapped to the graph, and that aren't in the graph to begin with, won't be called. You _could_ call an SV from the reads if the reads were long enough to just map over a 75 bp deletion and pay for its score cost, but they probably won't be. We don't do any local assembly or anything to try and recover novel SVs.
> 
> Does that help?

Thanks for your reply! Actucally I do not know the differents between the regular variant calling and vg deconstruct. As you said, the variants are in the graph, If I build the graph (cactus) by 10 high quality assemblies of samples and I just want the variants among them. From the manual, I think I can:
1. use vg deconstruct to get the vcf
2. map the illumina reads of each sample to the garph and call variants (giraffe/pack/call)
Is any differences bwtween them?

Also for new samples, If I augment it to the graph, could I just use vg deconstruct to call the variants?

Thank you!

**Comment by glennhickey** [comment]
`vg deconstruct` gives you all the variants in the graph.  `vg call`determines a subset of variants in the graph that are supported by the reads (aka genotyping).  

If you have a very simple graph, you can expect variants from `augment` to end up in your VCF from `deconstruct`.  But in practice, you generally need haplotypes in a GBWT in order to get a useful VCF from `deconstruct`.  In this case, you won't be able to get your variants from `augment` into the VCF.

**Comment by JD12138** [comment]
> `vg deconstruct` gives you all the variants in the graph. `vg call`determines a subset of variants in the graph that are supported by the reads (aka genotyping).
> 
> If you have a very simple graph, you can expect variants from `augment` to end up in your VCF from `deconstruct`. But in practice, you generally need haplotypes in a GBWT in order to get a useful VCF from `deconstruct`. In this case, you won't be able to get your variants from `augment` into the VCF.

hello,
So if i use your HGSVC graph in your science paper,then map my own reads to your graph.Then I use vg augment to embed my  gam to your graph.Then I use vg call to get the vcf file. Can I get novel variants which are not in the HGSVC graph?

And does vg call only genotype the variants already in the graph and not get new variants like GATK pipeline does?

**Comment by adamnovak** [comment]
> So if i use your HGSVC graph in your science paper,then map my own reads to your graph.Then I use vg augment to embed my gam to your graph.Then I use vg call to get the vcf file. Can I get novel variants which are not in the HGSVC graph?

Yes. But only if the novel variant is small enough for individual reads to show it.

> And does vg call only genotype the variants already in the graph and not get new variants like GATK pipeline does?

Yes. If you want new variants you are supposed to use `vg call` with `vg augment`.

**Comment by glennhickey** [comment]
`vg call` will report all variants in the graph.  If run on the augmented graph, it will be unaware of what's new and what's in the original graph and will report everything. 

The results are identical when using `vg call <graph>` and `vg snarls <graph> > snarls; vg call <graph> -r snarls`.  The second way just lets you re-use the snarls for other commands. 

 See the comment [above](https://github.com/vgteam/vg/issues/3567#issuecomment-1069367498) as to why `deconstruct` probably won't work on an augmented graph.

**Comment by jiadong324** [comment]
@glennhickey 

Thanks for your quick response! I still have the following questions:

1. I mapped short-reads of Sample-A to the graph provided in the Science paper. This graph contains ~167,000 SVs. Following the ```augment```, ```pack``` and ```call``` pipeline, I obtained around 1,300 SVs (insertions and deletions larger than 50bp) and a great number of SNPs and INDELs. So, I think the 1,300 SVs were newly add events to the graph after I ```map``` and ```augment```.

2. ```vg pack``` calculates the read support of the augmented graph. Dose this compute the number of read that support new links observed in the augmented graph. 

For example, if the original graph contains three nodes A, B and C, and the link is A->B->C. After augmentation, a new link is added from A to C. Then, ```vg pack``` computes the support of link A->C.

Thanks!
